enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+      iterationDomains.push_back(domain);
+      shapes.push_back(-1);
+    }
+
+    SmallVector<Value> inputs, outputs;
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<StringRef> iteratorTypes{linalgOp.getNumLoops(),
+                                         getParallelIteratorTypeName()};
+
+    for (OpOperand *output : linalgOp.getOutputOperands()) {
+      if (gutils->hasInvertPointer(output->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(output));
+        Value out = gutils->invertPointerM(output->get(), builder);
+        Value view = invertMemref(out, builder, op->getLoc());
+        outputs.push_back(view);
+      }
+    }
+
+    for (OpOperand *input : linalgOp.getInputOperands()) {
+      if (gutils->hasInvertPointer(input->get())) {
+        indexingMaps.push_back(linalgOp.getMatchingIndexingMap(input));
+        Value inp = gutils->invertPointerM(input->get(), builder);
+        Value view = invertMemref(inp, builder, op->getLoc());
+        inputs.push_back(view);
+      }
+    }
+
+    linalg::GenericOp adjoint = builder.create<linalg::GenericOp>(
+        op->getLoc(), outputs, inputs, indexingMaps, iteratorTypes);
+
+    int numInputs = inputs.size();
+    std::function<buildReturnFunction> buildFuncReturnOp =
+        [numInputs](OpBuilder &builder, Location loc,
+                    SmallVector<Value> retargs) {
+          builder.create<linalg::YieldOp>(
+              loc, ValueRange{retargs}.take_front(numInputs));
+          return;
+        };
+
+    Region *newOpRegion = newOp.getBlock()->getParent();
+    int numInputsNewOp = cast<linalg::GenericOp>(newOp).getInputs().size();
+    Region *adjointRegion = &adjoint.getBodyRegion();
+    int numInputsAdjoint = adjoint.getInputs().size();
+    Location loc = op->getLoc();
+    int numCaches = 0;
+    SmallVector<Value> pushCaches;
+
+    // std::function<std::pair<Value, Value>(Type)> hook = nullptr;
+    std::function<std::pair<Value, Value>(Type)> hook =
+        [newOpRegion, adjointRegion, loc, &numCaches = numCaches,
+         numInputsNewOp, numInputsAdjoint, &pushCaches = pushCaches](Type t) {
+          OpBuilder builder(newOpRegion);
+          Value pushCache = builder.create<enzyme::InitOp>(loc, t);
+          pushCaches.push_back(pushCache);
+          newOpRegion->addArgument(t, loc);
+
+          Value popCache = adjointRegion->insertArgument(
+              numInputsAdjoint + numCaches, t, loc);
+          numCaches++;
+          return std::pair<Value, Value>(pushCache, popCache);
+        };
+
+    gutils->Logic.differentiate(
+        gutils, *linalgOp.getBlock()->getParent(), adjoint.getBodyRegion(),
+        /*parentRegion=*/false, buildFuncReturnOp, hook);
+
+    // TODO add pushCaches to the yield in newOp
+    auto newOpYield = cast<linalg::YieldOp>(
+        cast<linalg::GenericOp>(newOp).getBodyRegion().front().getTerminator());
+    for (auto pc : pushCaches) {
```

Comment:
MLIR uses `auto` when the type is obvious from context (e.g., the RHS is a cast or a build) or difficult/impossible to spell (iterators, lambdas). Spell out the type in all the other cases as it will likely improve readability.